zbus HLP priority boost

rodrigopex opened the pull request:
> [!NOTE]
> **TL;DR:** Replace mutexes with semaphores to lock channels and implement the [Highest Locker Protocol (HLP)](https://www.geeksforgeeks.org/highest-locker-protocol-hlp/) priority boost for the zbus operations. The presentation about the feature is [here](https://docs.google.com/presentation/d/10ZpADOf4wunOvT1Xt_tk6PFM0Y-uYrN9wV5JdmCkgrM/edit?usp=sharing)

### ToDo :heavy_check_mark: 
The ~~DNM label is present~~ because there is some work to be done. Missing development actions from my side before merging:

- [x] Fix code documentation since we replaced mutexes with semaphores
- [x] Improve Kconfig help about the priority boost
- [ ] ~~Consider the runtime observer on the HOP calculation (only static observers are being considered)~~
- [ ] ~~Add the hop update for a channel when a runtime observer is added or removed from that~~ **_Those were removed because I plan to deprecate runtime observers_**. 


### Introduction
The current version of zbus offers message subscribers that can react immediately when the VDED transmits a message to them. So, despite zbus using mutexes, we can face some priority inversion.

For example, suppose four threads (T1, MS1, MS2, and S1) in the application. The highest priority is S1, and the lowest is T1 (see the figure below). Imagine MS1, MS2, and S1 are observing (waiting) a channel CHANNEL_A. The CHANNEL_A definition code is as follows:

```c
ZBUS_CHAN_DEFINE(channel_a,
		 struct msg,
		 NULL,			    
		 NULL,
		 ZBUS_OBSERVERS(L1, L2, MS1, MS2, S1),
		 ZBUS_MSG_INIT(0)
);
```

If T1 starts publishing (VDED process) to channel CHANNEL_A, any other thread with a priority higher than T1's can preempt the VDED, even if it has a priority lower than MS1, MS2, and S1. Even If no preemption happens, thread S1 will wait for MS1 and MS2 to receive and react to the message. The VDED execution is preempted more than once, which can increase the bus latency. 

![Current version](https://github.com/zephyrproject-rtos/zephyr/assets/35280/84e3f61c-94dc-4d77-8887-36f21505d9b9)

Another point is that zbus operations utilize mutexes to avoid race conditions when manipulating channels. However, as mutexes cannot operate within ISRs, zbus cannot either. That causes a lot of rework and makes zbus hard to use in several scenarios. 

This commit replaces the mutex usage with semaphores, so zbus loses the priority inheritance feature. To fill this gap, it implements the [HLP](https://www.geeksforgeeks.org/highest-locker-protocol-hlp/), which boosts the zbus publisher thread context based on the highest channels' observers' priority. And a good side effect is that zbus can run inside ISRs.

### Feature description

So, in addition to replacing mutexes with semaphores, zbus implements the Highest Locker Protocol that relies on the observers' thread priority to determine a temporary publisher priority. 

For example, suppose again that the application has four threads (T1, MS1, MS2, and S1). The highest priority is S1, and the lowest is T1 (see the figure below). Imagine that MS1, MS2, and S1 are observing (waiting) a channel CHANNEL_A. The figure below illustrates the VDED using the [HLP](https://www.geeksforgeeks.org/highest-locker-protocol-hlp/) algorithm.

![HLP](https://github.com/zephyrproject-rtos/zephyr/assets/35280/137ea0f8-4bc0-42b0-b21f-244959a94609)

### What are the benefits of the feature?

* The [HLP](https://www.geeksforgeeks.org/highest-locker-protocol-hlp/) is more effective for zbus than the mutexes priority inheritance;
* _**No broken APIs. 100% of compatibility.**_ Internal changes only;
* No bounded priority inversion will happen among the publisher and the observers;
* No other threads (that are not involved in the communication) with priority between T1 and S1 can preempt T1, avoiding unbounded priority inversion;
* Message subscribers will wait for the VDED to finish the message delivery process. So the VDED execution will be faster and more consistent;
* The HLP priority is dynamic and can change in execution;
* ZBus operations can be used inside ISRs;
* The priority boosting feature can be turned off, and plain semaphores can be used as the channel lock mechanism;
* HLP's major disadvantage, the Inheritance-related Priority Inversion, is acceptable in the zbus scenario since it will ensure a small bus latency.

rodrigopex committed:
zbus: add priority boost feature

Replace mutexes with semaphores to protect the channels in conjunction with
a priority boost algorithm based on the observers' priority.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex committed:
samples: zbus: fix samples to work with priority boost

Adjust the sample to work with the priority boost. It illustrates zbus
being used inside an ISR instead of into the main function.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex committed:
tests: zbus: fix unit tests to work with priority boost

The test considers using zbus inside ISR.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

andyross commented on subsys/zbus/zbus.c:
```diff
@@ -191,21 +194,133 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
+
+		if ((obs_prio_val > 0) && (chan_high_obs_prio_val > obs_prio_val)) {
+			atomic_cas(&chan->data->highest_observer_priority, chan_high_obs_prio_val,
+				   obs_prio_val);
```
I'm a little confused at this usage: atomic_cas() can fail if it races on the old value, and this isn't handling that case?  Is that because it can't fail by construction?  If so, you'd probably be better off using regular locking and normal variables instead of atomics.  If not, this needs to be fixed.

andyross replied:
This still needs a fix.  To clarify from earlier: atomic_cas() will swap the two priority values if and only if the old value still stored in the atomic variable has not changed from "obs_prio_val".  If it has, then atomic_cas() returns a failure condition that needs to be handled (usually with a retry of the get/compute/set sequence) by the caller.

The only way this isn't a bug is if it's impossible by construction that two threads can race on the highest_observer_priority field.  But I don't see any locking anywhere that would prevent that.

andyross commented on subsys/zbus/zbus.c:
```diff
@@ -191,21 +194,133 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
+
+			int p = chan_high_obs_prio_val - 1;
+
+			k_thread_priority_set(k_current_get(), MAX(p, 0));
```
I don't see any locking here?  The read of chan_high_obs_prio_val reflects the state of other blocked threads, but that can change between the line above where "p" is computed and here (e.g. someone might have aborted that thread, someone higher priority might have come in and blocked, etc...).

rodrigopex replied:
@andyross, I am waiting for the release to get back to that and tell you and others the intent here, OK?

rodrigopex replied:
@andyross I would like to get back on this feature. Would you happen to have time to talk about that next week?

andyross reviewed (commented):
Some worries about the synchronization strategy after a (very brief) look.  Broadly: the decisions made here need to be made based on the behavior of  multiple async threads interacting with the bus, it's not just a produce/consume thing which is normally the regime of lockless algorithms.

IMHO you're really going to struggle to do this correctly and lockless using only atomics (I won't say such a thing can't be done, but I will say I'd expect it to be the subject of a published paper or three, heh).

Really, this wants a per-bus spinlock where you can check the state with cleaner rules and make atomic decisions using normal logic.

andyross commented on subsys/zbus/zbus.c:
```diff
@@ -191,21 +194,133 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
+			int p = chan_high_obs_prio_val - 1;
+
+			k_thread_priority_set(k_current_get(), MAX(p, 0));
+
```
There is still a race condition here.  "chan_high_obs_prio_val" was read above (at line 248), and is being used here to compute "p".  But between those two lines of code another thread may have come along and changed highest_observer_priority, such that the p value you are using here is now wrong (i.e. not the highest observer priority).

andyross reviewed (changes requested):
Refresh/elaborate the synchronization comments.  Again, really I think this lockless design is going to be a dangerous approach vs. just putting a spinlock around the bus priority computation.

Also, it might be worth thinking of ways to abstract out the priority elevation code such that this can share it with k_mutex.  The problem is basically identical.

rodrigopex reviewed (commented):
@andyross, I've added the spinlock to the critical code portion and made some changes to the code. Please take a look.

andyross commented on subsys/zbus/zbus.c:
```diff
@@ -191,21 +196,142 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
+
+	if (err) {
+		if (boosting) {
+			LOG_ERR("Something went wrong. Restoring %s priority", _THREAD_NAME);
```
LOG_ERR() is almost always going to appear in the log stream, and "Something went wrong" isn't informative enough to be worth logging, IMHO. (FWIW: k_sem_take is only going to return non-zero on a timeout, or plausibly something like someone directly waking the thread up or destroying the semaphore).    The function has its own return value, you can just return that and let the caller log whatever the broken state is.

andyross commented on subsys/zbus/zbus.c:
```diff
@@ -191,21 +196,142 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
+
+	int current_thread_priority = k_thread_priority_get(k_current_get());
+
+	if (unlikely(obs->data->priority != current_thread_priority)) {
```
This test wants to be inside the lock too.  Basically anything that touches the shared state in obs->data is going to want to be locked absent very careful analysis about what races are possible.

andyross replied:
Also style: I'm not personally a fan of the likely/unlikely macros in situations where you aren't deliberately trying to generate specific code sequences.  The overwhelming majority of devices Zephyr is deployed on don't have large branch prediction penalties, and in the modern world CPUs which do tend to have branch predictors that are smarter than rando C programmers like us anyway. :)

rodrigopex replied:
> This test wants to be inside the lock too. Basically anything that touches the shared state in obs->data is going to want to be locked absent very careful analysis about what races are possible.

Only a single thread can change an observer's priority at a time, as an observer is "attached" to a thread. It is not expected for the `_zbus_obs_attach_to_thread` to be called more than once concurrently, so changing the observer priority without spinlock would be safe.

andyross replied:
> Only a single thread can change an observer's priority at a time

Is that enforced or just by convention?  Could an attacker arrange things such that two threads race in this block here and end up doing incorrect things?

andyross commented on subsys/zbus/zbus.c:
```diff
@@ -366,8 +506,22 @@ int zbus_obs_is_chan_notification_masked(const struct zbus_observer *obs,
+
+	obs->data->enabled = enabled;
+
+	_zbus_update_all_channels_hop(obs);
```
Similarly wants locking here: what happens if someone crosses between the enabled set and the update?

rodrigopex replied:
All the enable/disable actions finish with an update. If someone crosses an update process and sets the enable over again, a second update will happen just after the first one finishes, and the zbus will consider the latest value during the channel's HOP (Highest Observer Priority). The `_zbus_update_all_channels_hop` will lock the channel access to the `highest_observer_priority` field. So, the sensitive part is spinlocked. 

Am I being superficial in my analyses of the situation?

rodrigopex replied:
@andyross, please take a look at the `_zbus_chan_update_hop` that is the heart of the HLP protocol. The `_zbus_update_all_channels_hop(obs)` calls the `_zbus_chan_update_hop` for each channel that has `obs` as an observer.

rodrigopex replied:
Should the channel update part be more protected, or is it okay already?

andyross reviewed (commented):
This looks much cleaner to me, thanks.

Found a few spots that either want to be inside the lock or to have a comment explaining why they don't need it.  I won't promise those are the only ones though, just the ones I saw. :)

Also some misc. nitpicks that aren't worth a -1.

rodrigopex committed:
tests: zbus: add HLP priority boost tests

Add the dedicated tests for HLP priority boost.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex committed:
samples: zbus: add priority boost sample to zbus

This sample illustrates how to properly use zbus priority boost feature.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex reviewed (commented):
@andyross

I have solved unsafe code with spinlocks. All the actions that affect observers and channels (priority-related fields) have a specific lock. With that, I established two-level locking steps, where the developer will cause the `_zbus_obs_slock` by attaching the observer to a thread or by disabling it or some of its observations. The observer priority change will make the bus lock the channels with spin locking the `_zbus_chan_slock` and update the highest priority observers for all channels with that observer. That seems to be safe now. Please take a look and let me know. 

> [!WARNING]
> I kept the **DNM** label because I must add specific tests for the `priority_boost` feature. I did not add that yet because I need to make the implementation stable enough to spend time testing it.

andyross reviewed (dismissed):
Synchronization looks good to me.  Feature seems very reasonable.

It's worth noting that this has the same hole that (IIRC) k_mutex does where if you abort a blocked thread that has "granted" its priority to the owner, there's no hook in the system to recover and the owner keeps that priority until it exits the boosted state.  This is semi-benign, but something we occasionally talk about fixing in mutex.

rodrigopex committed:
doc: zbus: add priority boost documentation

Add priority boost documentation. Replace the ISR limitation since it is
not valid anymore.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex commented:
The DNM label is present because there is some work to be done. Missing development actions from my side before merging:

- [ ] Fix code documentation since we replaced mutexes with semaphores
- [ ] Consider the runtime observer on the HOP calculation (only static observers are being considered)
- [ ] Add the hop update for a channel when a runtime observer is added to that

fabiobaltieri commented on include/zephyr/zbus/zbus.h:
```diff
@@ -348,10 +370,13 @@ struct zbus_channel_observation {
 #define ZBUS_SUBSCRIBER_DEFINE_WITH_ENABLE(_name, _queue_size, _enable)                            \
 	K_MSGQ_DEFINE(_zbus_observer_queue_##_name, sizeof(const struct zbus_channel *),           \
 		      _queue_size, sizeof(const struct zbus_channel *));                           \
+	static struct zbus_observer_data _CONCAT(_zbus_obs_data_, _name) = {                       \
+		.enabled = _enable IF_ENABLED(CONFIG_ZBUS_PRIORITY_BOOST,                          \
+					      (, .priority = _MIN_THREAD_PRIORITY))};              \
```
```c
        static struct zbus_observer_data _CONCAT(_zbus_obs_data_, _name) = {
                .enabled = _enable, 
                IF_ENABLED(CONFIG_ZBUS_PRIORITY_BOOST, ( 
                .priority = _MIN_THREAD_PRIORITY,
                ))
        };
```

fabiobaltieri commented on include/zephyr/zbus/zbus.h:
```diff
@@ -348,10 +370,13 @@ struct zbus_channel_observation {
 		ZBUS_OBSERVER_NAME_INIT(_name) /* Name field */                                    \
 			.type = ZBUS_OBSERVER_SUBSCRIBER_TYPE,                                     \
-		.enabled = _enable, .queue = &_zbus_observer_queue_##_name}
+		.data = &_CONCAT(_zbus_obs_data_, _name), .queue = &_zbus_observer_queue_##_name}
```
```c
        STRUCT_SECTION_ITERABLE(zbus_observer, _name) = {
                ZBUS_OBSERVER_NAME_INIT(_name) /* Name field */
                .type = ZBUS_OBSERVER_SUBSCRIBER_TYPE,
                .data = &_CONCAT(_zbus_obs_data_, _name), 
                .queue = &_zbus_observer_queue_##_name,
        }
```

fabiobaltieri commented on include/zephyr/zbus/zbus.h:
```diff
@@ -378,10 +403,13 @@ struct zbus_channel_observation {
+	STRUCT_SECTION_ITERABLE(zbus_observer, _name) = {                                          \
+		ZBUS_OBSERVER_NAME_INIT(_name) /* Name field */                                    \
+			.type = ZBUS_OBSERVER_LISTENER_TYPE,                                       \
+		.data = &_CONCAT(_zbus_obs_data_, _name), .callback = (_cb)}
```
```c
        static struct zbus_observer_data _CONCAT(_zbus_obs_data_, _name) = {
                .enabled = _enable, 
                IF_ENABLED(CONFIG_ZBUS_PRIORITY_BOOST, (
                .priority = _MIN_THREAD_PRIORITY,
                ))
        };
        STRUCT_SECTION_ITERABLE(zbus_observer, _name) = {
                ZBUS_OBSERVER_NAME_INIT(_name) /* Name field */
                .type = ZBUS_OBSERVER_LISTENER_TYPE,
                .data = &_CONCAT(_zbus_obs_data_, _name), 
                .callback = (_cb),
        }
```

fabiobaltieri commented on include/zephyr/zbus/zbus.h:
```diff
@@ -406,10 +434,13 @@ struct zbus_channel_observation {
  */
 #define ZBUS_MSG_SUBSCRIBER_DEFINE_WITH_ENABLE(_name, _enable)                                     \
 	static K_FIFO_DEFINE(_zbus_observer_fifo_##_name);                                         \
+	static struct zbus_observer_data _CONCAT(_zbus_obs_data_, _name) = {                       \
```
you got the idea :-)

fabiobaltieri commented on subsys/zbus/zbus.c:
```diff
@@ -12,6 +12,16 @@
 #include <zephyr/zbus/zbus.h>
 LOG_MODULE_REGISTER(zbus, CONFIG_ZBUS_LOG_LEVEL);
 
+#define _THREAD_NAME                                                                               \
+	(IS_ENABLED(CONFIG_THREAD_NAME) ? k_thread_name_get(k_current_get()) : "publisher")
```
```c
#if CONFIG_THREAD_NAME
#define ZBUS_THREAD_NAME k_thread_name_get(k_current_get())
#else
#define ZBUS_THREAD_NAME "publisher"
#endif
```

I think _ prefix deifnes should be avoided, `IS_ENABLED` is a bit of an odd thing, if normal ifdef works just use those

rodrigopex replied:
Removed.

fabiobaltieri commented on subsys/zbus/zbus.c:
```diff
@@ -191,21 +201,175 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
+
+	int chan_highest_observer_priority = _MIN_THREAD_PRIORITY;
+
+	K_SPINLOCK(&_zbus_chan_slock)
+	{
+		for (int16_t i = chan->data->observers_start_idx,
+			     limit = chan->data->observers_end_idx;
+		     i < limit; ++i) {
```
```c
        uint16_t start_idx = chan->data->observers_start_idx;
        uint16_t end_idx = chan->data->observers_end_idx;
        K_SPINLOCK(&_zbus_chan_slock) {
                for (int16_t i = start_idx, limit = end_idx; i < limit; ++i) {
```

fabiobaltieri commented on subsys/zbus/zbus.c:
```diff
@@ -191,21 +201,175 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
 
+#if defined(CONFIG_ZBUS_PRIORITY_BOOST)
+
+static inline void _zbus_chan_update_hop(const struct zbus_channel *chan)
```
don't use `_` prefix functions please, it's reserved for the library or something IIRC

fabiobaltieri commented on subsys/zbus/zbus.c:
```diff
@@ -191,21 +201,175 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
+
+	int current_thread_priority = k_thread_priority_get(k_current_get());
+
+	K_SPINLOCK(&_zbus_obs_slock)
+	{
```
`K_SPINLOCK(&_zbus_obs_slock) {`

fabiobaltieri commented on subsys/zbus/zbus.c:
```diff
@@ -191,21 +201,175 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
+	_ZBUS_ASSERT(!k_is_in_isr(), "cannot detach from an ISR");
+	_ZBUS_ASSERT(obs != NULL, "obs is required");
+
+	K_SPINLOCK(&_zbus_obs_slock)
+	{
```
`K_SPINLOCK(&_zbus_obs_slock) {`

fabiobaltieri commented on subsys/zbus/zbus.c:
```diff
@@ -191,21 +201,175 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
+	if (!k_is_in_isr()) {
+		*prio = k_thread_priority_get(k_current_get());
+
+		K_SPINLOCK(&_zbus_chan_slock)
+		{
```
ditto

fabiobaltieri commented on subsys/zbus/zbus.c:
```diff
@@ -330,22 +504,36 @@ int zbus_obs_set_chan_notification_mask(const struct zbus_observer *obs,
-	     i < limit; ++i) {
-		STRUCT_SECTION_GET(zbus_channel_observation, i, &observation);
-		STRUCT_SECTION_GET(zbus_channel_observation_mask, i, &observation_mask);
+	K_SPINLOCK(&_zbus_obs_slock)
+	{
+		for (int16_t i = chan->data->observers_start_idx,
+			     limit = chan->data->observers_end_idx;
+		     i < limit; ++i) {
```
same as the other one

fabiobaltieri commented on subsys/zbus/zbus.c:
```diff
@@ -354,20 +542,46 @@ int zbus_obs_is_chan_notification_masked(const struct zbus_observer *obs,
-	     i < limit; ++i) {
-		STRUCT_SECTION_GET(zbus_channel_observation, i, &observation);
-		STRUCT_SECTION_GET(zbus_channel_observation_mask, i, &observation_mask);
+	K_SPINLOCK(&_zbus_obs_slock)
+	{
+		for (int16_t i = chan->data->observers_start_idx,
+			     limit = chan->data->observers_end_idx;
+		     i < limit; ++i) {
```
same

fabiobaltieri commented on subsys/zbus/zbus.c:
```diff
@@ -354,20 +542,46 @@ int zbus_obs_is_chan_notification_masked(const struct zbus_observer *obs,
+{
+	_ZBUS_ASSERT(obs != NULL, "obs is required");
+
+	K_SPINLOCK(&_zbus_obs_slock)
+	{
```
same

fabiobaltieri commented on tests/subsys/zbus/unittests/src/main.c:
```diff
@@ -760,4 +770,34 @@ ZTEST(basic, test_specification_based__zbus_sub_wait_msg)
+{
+	ztest_test_skip();
+}
+
```
nit: drop, symmetry...

rodrigopex replied:
I am following the documentation. Please change that to avoid confusion: https://docs.zephyrproject.org/latest/develop/test/ztest.html#skipping-tests

fabiobaltieri replied:
I'm talking about the empty line, you have none separating the #if and #else but you left one on the #endif

fabiobaltieri replied:
the doc has a good example though can you skip with `Z_TEST_SKIP_IFDEF` rather than building this out?

rodrigopex replied:
I would like to skip it if it is not defined. Should I "not" that?

fabiobaltieri replied:
of course, can you use `Z_TEST_SKIP_IFNDEF`? juts keeps things a bit neater and make sure the whole code gets compiled regarless of the config, you obviously can't do that if something you use in the test is in a conditional itself but I don't think it's the case here? If it's the case then fine as it is just drop the blank line. :-)

Same logic as #ifdef versus IS_ENABLED.

rodrigopex replied:
I did not know there was such a thing. I will use that. Thanks.

fabiobaltieri replied:
yeah me neither until you linked that doc :-)

rodrigopex replied:
I see why I needed the ifdef style. The `zbus_obs_attach_to_thread` only exists when the priority boost is enabled.

fabiobaltieri replied:
ok then fine as it is, just drop that empty line

fabiobaltieri replied:
well I would only ifdef/endif on the content and not duplicate the `ZTEST(basic, test_specification_based__zbus_obs_attach_detach)` but up to you

fabiobaltieri commented on tests/subsys/zbus/unittests/src/main.c:
```diff
@@ -700,10 +706,14 @@ ZTEST(basic, test_specification_based__zbus_obs_set_chan_notification_mask)
 ZBUS_SUBSCRIBER_DEFINE(foo_sub, 1);
 
+static struct zbus_observer_data _zbus_obs_data_invalid_sub = {
+	.enabled =
+		false IF_ENABLED(CONFIG_ZBUS_PRIORITY_BOOST,
+				 (, .priority = ATOMIC_INIT(CONFIG_NUM_PREEMPT_PRIORITIES - 1)))};
 STRUCT_SECTION_ITERABLE(zbus_observer,
 			invalid_sub) = {ZBUS_OBSERVER_NAME_INIT(invalid_sub) /* Name field */
 						.type = ZBUS_OBSERVER_SUBSCRIBER_TYPE,
-					.enabled = false, .queue = NULL};
+					.data = &_zbus_obs_data_invalid_sub, .queue = NULL};
```
same as in the previous ones

fabiobaltieri commented on tests/subsys/zbus/unittests/src/main.c:
```diff
@@ -223,11 +223,16 @@ static void wq_dh_cb(struct k_work *item)
 ZBUS_MSG_SUBSCRIBER_DEFINE_WITH_ENABLE(foo2_msg_sub, false);
+
 static K_FIFO_DEFINE(_zbus_observer_fifo_invalid_obs);
+static struct zbus_observer_data _zbus_obs_data_invalid_obs = {
+	.enabled =
+		false IF_ENABLED(CONFIG_ZBUS_PRIORITY_BOOST,
+				 (, .priority = ATOMIC_INIT(CONFIG_NUM_PREEMPT_PRIORITIES - 1)))};
```
I'll leave this to you :-)

fabiobaltieri commented on tests/subsys/zbus/hlp_priority_boost/src/main.c:
```diff
@@ -0,0 +1,180 @@
+
+static struct msg_testing_01 msg = {.seq = 0};
+
+void publisher_thread(void *ptr1, void *ptr2, void *ptr3)
```
static

fabiobaltieri commented on tests/subsys/zbus/hlp_priority_boost/src/main.c:
```diff
@@ -0,0 +1,180 @@
+
+ZBUS_SUBSCRIBER_DEFINE(sub1, 4);
+
+void consumer_msg_sub_thread(void *ptr1, void *ptr2, void *ptr3)
```
static

fabiobaltieri commented on tests/subsys/zbus/hlp_priority_boost/src/main.c:
```diff
@@ -0,0 +1,180 @@
+		 ZBUS_MSG_INIT(0)                   /* Initial value major 0, minor 1, build 1023 */
+);
+
+void consumer_sub_thread(void *ptr1, void *ptr2, void *ptr3)
```
static

fabiobaltieri commented on tests/subsys/zbus/hlp_priority_boost/src/main.c:
```diff
@@ -0,0 +1,180 @@
+
+ZBUS_MSG_SUBSCRIBER_DEFINE(msub1);
+
+K_SEM_DEFINE(sync_sem, 1, 1);
+K_SEM_DEFINE(done_sem, 0, 1);
```
static

fabiobaltieri commented on samples/subsys/zbus/priority_boost/src/main.c:
```diff
@@ -0,0 +1,153 @@
+}
+ZBUS_LISTENER_DEFINE(l2, l2_callback);
+
+int t1_thread(void *ptr1, void *ptr2, void *ptr3)
```
static, I'll leave some more here for the rest of the file: static static static static static static static  just spread them around :-)

fabiobaltieri commented on samples/subsys/zbus/priority_boost/src/main.c:
```diff
@@ -0,0 +1,153 @@
+		}
+
+		/* Faking some workload */
+		k_busy_wait(200000);
```
is this 20000 or 2000000? `200 * USEC_PER_MSEC` :-)

fabiobaltieri reviewed (changes requested):
Just a round of cosmetic/coding style comments. You are trying to run clang-format on these right? I think you should avoid it altogether for macro heavy files.

rodrigopex reviewed (commented):
@fabiobaltieri changes requested attended.

rodrigopex commented:
@fabiobaltieri and others, I have created another PR adding the `K_SPINLOCK` to the `FOR_EACH` indent rule. #66982

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -154,6 +171,8 @@ struct zbus_channel_observation {
 #define _ZBUS_CPP_EXTERN
 #endif /* __cplusplus */
 
+#define _MIN_THREAD_PRIORITY (CONFIG_NUM_PREEMPT_PRIORITIES - 1)
```
To improve readability, I usually use, in my personal code, the `_` in front of the local scope variables. I will remove that.

rodrigopex commented on include/zephyr/zbus/zbus.h:
```diff
@@ -345,13 +380,25 @@ struct zbus_channel_observation {
-		.enabled = _enable, .queue = &_zbus_observer_queue_##_name}
+#define ZBUS_SUBSCRIBER_DEFINE_WITH_ENABLE(_name, _queue_size, _enable)       \
+	K_MSGQ_DEFINE(_zbus_observer_queue_##_name,                           \
+		sizeof(const struct zbus_channel *),                          \
```
It would force breaking the lines. As I am not using the clang-format, I do not see an issue in letting them like this. Do you? I will start facing many compliance issues because you feel uncomfortable with clang-format results for macros. I suggest fixing the clang-format to avoid this kind of problem.

rodrigopex replied:
> You sure? Looks to me it'd fit alright
> 
> ```c
>         K_MSGQ_DEFINE(_zbus_observer_queue_##_name,                           \
>                       sizeof(const struct zbus_channel *),                    \
>                       _queue_size,                                            \
>                       sizeof(const struct zbus_channel *)                     \
>         );        
> ```
> 
> it's also usually common practice to either do all arguments in line or one per line (or sometimes split them logically if there's an obvious pairing or something), but the rest of the file is already off that so I normally would not point that out.
> 
> Also I don't think that the parenthesis alignment is strictly part of the coding guideline, but I do care about the indentation style being coherent within the file, and that's how the rest of the file does it (which so happen to be coherent with most of the code base).

I thought you were mentioning aligning with `ZBUS_SUBSCRIBER_DEFINE_WITH_ENABLE( `. That part is ok to go. Sorry, sometimes I get confused with the request change point.

rodrigopex replied:
Done.

fabiobaltieri replied:
Thanks, unfortunately the way github show comment context makes it confusing to understand if the comment was on just a line or the whole block. Starting to do more snippets to avoid ambiguity these days but that takes even more time.

Anyway, looks good, thanks for your patience with this, will do another pass later.

rodrigopex commented on subsys/zbus/zbus.c:
```diff
@@ -191,21 +198,171 @@ static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t
+
+			const struct zbus_observer *obs = observation->obs;
+
+			if (!obs->data->enabled || observation_mask->enabled) {
```
That is correct.

rodrigopex reviewed (commented):
@fabiobaltieri, requests attended.

rodrigopex commented:
> @fabiobaltieri, requests attended.

I hope it is fine now. @fabiobaltieri

andyross reviewed (approved):
refresh +1
